components/application/application.js

```diff
@@ -0,0 +1,204 @@
+            description  : req.body.description,
+            requestedBy  : req.auth.id,
+            status       : "pending"
+        })
+
+        // client made an invalid request
+        let err = application.validateSync()
+        if(err){
+            return res.status(400).json({
+                err : err.message
+            })
+        }
+
+        await application.save()
+
+        res.status(200).json({
+            msg  : "Application received.",
+        })
+    }catch(e){
+        return next(e)
+    }
+}
+
+
+async function getMyApplications(req,res,next){
+    try{
+        let applications = await Application.find({
+            requestedBy : req.auth.id
+        })
+        // there are some details which we cannot provide to the user
+        applications = applications.map( a => ({
+                // should we wend this one out?
+                id            : a.id,
+                amount        : a.amount,
+                payments      : a.payments,
+                paymentAmount : a.paymentAmount,
+                reason        : a.reason,
+                description   : a.description,
+                status        : a.status
+            })
+        )
+        res.status(200).json(applications)
+    }catch(e){
+        return next(e)
+    }
+}
+
+async function getViewApplication(req,res,next){
+    try{
+        // find application by id and user
+        let a = await Application.findOne({
+            _id         : req.params.id,
+            requestedBy : req.auth.id,
+        })
+        if(!a){
+            return next(
+                new Error("No such application for the current user.")
+            )
+        }
+
+        // we don't let the user see very single detail.
+        let result = {
+            amount        : a.amount,
+            payments      : a.payments,
+            paymentAmount : a.paymentAmount,
+            reason        : a.reason,
+            description   : a.description,
+            status        : a.status,
+            requestedAt   : a.createdAt
+        }
+
+        if("rejected" === a.status){
+            result.rejectedReason = a.rejectedReason
+        }
+
+        return res.status(200).json(result)
+    }catch(e){
+        return next(e)
+    }
+}
+
+/**
+ * Cancels an application. User initiated.
+ * When this happens, we need to give the user a "cooldown period" before he/she can apply again.
+ */
+async function postCancelApplication(req,res,next){
+    try{
+        let application = await Application.findOne({
+            // this is necessary.
+            _id         : req.params.id,
+            requestedBy : req.auth.id,
+            status      : "pending"
+        })
+
+        // did not find it.
+        if(application == null){
+            return next(
+                new Error("No pending application with that id for the current user.")
+            )
+        }
+
+        // change status to cancelled.
+        application.status = "cancelled"

```

Comment:
I think we should create an object to capture the different statuses? That way we can do something like APPLICATION_STATUS.cancelled, `.pending` etc? (Hi Typescript!) instead of writing the string literal?

Reply:
Good call. I'll create an object for the constants at the top of the file but maybe it should stay with the schema

Reply:
Done. It now resides together with the application schema